commit 3c7670d56ac650b988e560298a71e7f8d7a6828d
Author: Liran Alon <liran.alon@oracle.com>
Date:   Tue Nov 20 18:03:25 2018 +0200

    KVM: VMX: Update shared MSRs to be saved/restored on MSR_EFER.LMA changes
    
    [ Upstream commit f48b4711dd6e1cf282f9dfd159c14a305909c97c ]
    
    When guest transitions from/to long-mode by modifying MSR_EFER.LMA,
    the list of shared MSRs to be saved/restored on guest<->host
    transitions is updated (See vmx_set_efer() call to setup_msrs()).
    
    On every entry to guest, vcpu_enter_guest() calls
    vmx_prepare_switch_to_guest(). This function should also take care
    of setting the shared MSRs to be saved/restored. However, the
    function does nothing in case we are already running with loaded
    guest state (vmx->loaded_cpu_state != NULL).
    
    This means that even when guest modifies MSR_EFER.LMA which results
    in updating the list of shared MSRs, it isn't being taken into account
    by vmx_prepare_switch_to_guest() because it happens while we are
    running with loaded guest state.
    
    To fix above mentioned issue, add a flag to mark that the list of
    shared MSRs has been updated and modify vmx_prepare_switch_to_guest()
    to set shared MSRs when running with host state *OR* list of shared
    MSRs has been updated.
    
    Note that this issue was mistakenly introduced by commit
    678e315e78a7 ("KVM: vmx: add dedicated utility to access guest's
    kernel_gs_base") because previously vmx_set_efer() always called
    vmx_load_host_state() which resulted in vmx_prepare_switch_to_guest() to
    set shared MSRs.
    
    Fixes: 678e315e78a7 ("KVM: vmx: add dedicated utility to access guest's kernel_gs_base")
    Reported-by: Eyal Moscovici <eyal.moscovici@oracle.com>
    Reviewed-by: Mihai Carabas <mihai.carabas@oracle.com>
    Reviewed-by: Liam Merwick <liam.merwick@oracle.com>
    Reviewed-by: Jim Mattson <jmattson@google.com>
    Signed-off-by: Liran Alon <liran.alon@oracle.com>
    Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
    Signed-off-by: Sasha Levin <sashal@kernel.org>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index e55f7a90d4b2..b1aef90aab91 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -962,6 +962,7 @@ struct vcpu_vmx {
 	struct shared_msr_entry *guest_msrs;
 	int                   nmsrs;
 	int                   save_nmsrs;
+	bool                  guest_msrs_dirty;
 	unsigned long	      host_idt_base;
 #ifdef CONFIG_X86_64
 	u64 		      msr_host_kernel_gs_base;
@@ -2874,6 +2875,20 @@ static void vmx_prepare_switch_to_guest(struct kvm_vcpu *vcpu)
 
 	vmx->req_immediate_exit = false;
 
+	/*
+	 * Note that guest MSRs to be saved/restored can also be changed
+	 * when guest state is loaded. This happens when guest transitions
+	 * to/from long-mode by setting MSR_EFER.LMA.
+	 */
+	if (!vmx->loaded_cpu_state || vmx->guest_msrs_dirty) {
+		vmx->guest_msrs_dirty = false;
+		for (i = 0; i < vmx->save_nmsrs; ++i)
+			kvm_set_shared_msr(vmx->guest_msrs[i].index,
+					   vmx->guest_msrs[i].data,
+					   vmx->guest_msrs[i].mask);
+
+	}
+
 	if (vmx->loaded_cpu_state)
 		return;
 
@@ -2934,11 +2949,6 @@ static void vmx_prepare_switch_to_guest(struct kvm_vcpu *vcpu)
 		vmcs_writel(HOST_GS_BASE, gs_base);
 		host_state->gs_base = gs_base;
 	}
-
-	for (i = 0; i < vmx->save_nmsrs; ++i)
-		kvm_set_shared_msr(vmx->guest_msrs[i].index,
-				   vmx->guest_msrs[i].data,
-				   vmx->guest_msrs[i].mask);
 }
 
 static void vmx_prepare_switch_to_host(struct vcpu_vmx *vmx)
@@ -3418,6 +3428,7 @@ static void setup_msrs(struct vcpu_vmx *vmx)
 		move_msr_up(vmx, index, save_nmsrs++);
 
 	vmx->save_nmsrs = save_nmsrs;
+	vmx->guest_msrs_dirty = true;
 
 	if (cpu_has_vmx_msr_bitmap())
 		vmx_update_msr_bitmap(&vmx->vcpu);
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1096_3c7670d56ac650b988e560298a71e7f8d7a6828d.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1096_3c7670d56ac650b988e560298a71e7f8d7a6828d.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
